src/engines/arps/ARP/audio.cpp

```diff
@@ -1,185 +1,326 @@
-  {
-    if (rounds % 2 == 0) output.push_back(note + 12);
-    else output.push_back(note - 12);
-  }
 
   // Action Handlers //
   void Audio::action(itc::prop_change<&Props::playmode>, Playmode pm) noexcept
   {
-    switch(pm){
-    case Playmode::up: playmode_func_ = up; break;
-    case Playmode::down: playmode_func_ = down; break;
-    default: playmode_func_ = up; break;
+    using namespace play_modes;
+    switch (pm) {
+      case Playmode::up: playmode_func_ = up; break;
+      case Playmode::down: playmode_func_ = down; break;
+      case Playmode::updown: playmode_func_ = updown; break;
+      case Playmode::downup: playmode_func_ = downup; break;
+      case Playmode::updowninc: playmode_func_ = updowninc; break;
+      case Playmode::downupinc: playmode_func_ = downupinc; break;
+      case Playmode::manual: playmode_func_ = manual; break;
+      case Playmode::chord: playmode_func_ = chord; break;
+      default: playmode_func_ = play_modes::up; break;
     }
   }
-    
+
   void Audio::action(itc::prop_change<&Props::octavemode>, OctaveMode om) noexcept
   {
-    switch(om){
-    case OctaveMode::standard: octavemode_func_ = standard; break;
-    case OctaveMode::octaveupunison: octavemode_func_ = octaveupunison; break;
-    case OctaveMode::fifthunison: octavemode_func_ = fifthunison; break;
-    case OctaveMode::octaveup: octavemode_func_ = octaveup; break;
-    case OctaveMode::octaveupdown: octavemode_func_ = octaveupdown; break;
-    case OctaveMode::octavedownup: octavemode_func_ = octavedownup; break;
-    default: octavemode_func_ = standard; break;
+    using namespace octave_modes;
+    switch (om) {
+      case OctaveMode::standard: octavemode_func_ = standard; break;
+      case OctaveMode::octaveupunison: octavemode_func_ = octaveupunison; break;
+      case OctaveMode::fifthunison: octavemode_func_ = fifthunison; break;
+      case OctaveMode::octaveup: octavemode_func_ = octaveup; break;
+      case OctaveMode::octavedownup: octavemode_func_ = octavedownup; break;
+      default: octavemode_func_ = octave_modes::standard; break;
     }
   }
 
   void Audio::action(itc::prop_change<&Props::note_length>, float nl) noexcept
   {
     note_length_ = nl;
-    note_off_frames = (int)(nl * (float)_samples_per_beat);
+    note_off_frames = (int) (nl * (float) _samples_per_beat);
   }
 
   void Audio::action(itc::prop_change<&Props::subdivision>, int sd) noexcept
   {
     _samples_per_beat = _samples_per_quarternote / sd;
-    note_off_frames = (int)(note_length_ * (float)_samples_per_beat);
+    note_off_frames = (int) (note_length_ * (float) _samples_per_beat);
+    switch (sd) {
+      case 1: note = core::clock::notes::quarter; break;
+      case 2: note = core::clock::notes::eighth; break;
+      case 3: note = core::clock::notes::eighthtriplet; break;
+      case 4: note = core::clock::notes::sixteenth; break;
+      default: note = core::clock::notes::quarter; break;
```

Comment:
In this case, the default is fine, but use something like `default: OTTO_UNREACHABLE;` to mark that this code should never be reached. That should disable any warnings about something not being handled, and will give us an assertion error if that code is ever reached.